File: sampleCode/Python/utilities/json_helper.py
"""Translation between the API's JSON and the sample's Python models.

The Impact API speaks `camelCase`; Python is written in `snake_case`. These two
functions do the conversion, and `ApiModel` gives every model a tolerant
constructor so the samples keep working when the API starts returning a field they
have never heard of.

That tolerance is the important part. The API gains fields over time, and a model
that raises on an unexpected key turns a routine addition into a broken sample, so
unknown keys are kept in `extra` rather than rejected. Print `model.extra` when you
want to see everything a response really contained.
"""

import dataclasses
import enum
import re
import uuid
from typing import Any, Mapping, Self

# Matches the position just before an inner capital letter: the boundary between
# `hash` and `Id` in `hashId`.
_CAMEL_BOUNDARY = re.compile(r"(?<!^)(?=[A-Z])")


def camel_case(name: str) -> str:
    """Convert a Python field name to the API's spelling: `hash_id` -> `hashId`."""
    head, *rest = name.split("_")
    return head + "".join(word[:1].upper() + word[1:] for word in rest)


def snake_case(name: str) -> str:
    """Convert an API field name to Python's spelling: `hashId` -> `hash_id`."""
    return _CAMEL_BOUNDARY.sub("_", name).lower()


def to_json_value(value: Any) -> Any:
    """Render one Python value the way the API expects to receive it.

    Enums become their string value, UUIDs become their text form, models become
    dictionaries, and collections are converted item by item.
    """
    if isinstance(value, ApiModel):
        return value.to_api()
    if isinstance(value, enum.Enum):
        return value.value
    if isinstance(value, uuid.UUID):
        return str(value)
    if isinstance(value, (list, tuple)):
        return [to_json_value(item) for item in value]
    if isinstance(value, dict):
        return {camel_case(key): to_json_value(item) for key, item in value.items()}
    return value


@dataclasses.dataclass
class ApiModel:
    """Base class for every request and response model in the samples.

    Subclasses are plain dataclasses. This base adds `from_api`, which builds an
    instance from a JSON object without tripping over fields it does not know, and
    `to_api`, which turns an instance back into the JSON the API expects.
    """

    # Anything the API returned that this model does not declare. Nothing is
    # silently discarded: inspect `model.extra` to see what else came back.
    extra: dict[str, Any] = dataclasses.field(
        default_factory=dict, repr=False, compare=False, kw_only=True
    )

    @classmethod
    def from_api(cls, payload: Mapping[str, Any]) -> Self:
        """Build a model from one JSON object, keeping unknown fields in `extra`."""
        known = {f.name for f in dataclasses.fields(cls) if f.name != "extra"}
        arguments: dict[str, Any] = {}
        extra: dict[str, Any] = {}

        for api_name, value in payload.items():
            python_name = snake_case(api_name)
            if python_name in known:
                arguments[python_name] = value
            else:
                # A field the API has added, or one this sample has no use for.
                # Keep it rather than failing.
                extra[api_name] = value

        return cls(**arguments, extra=extra)

    @classmethod
    def list_from_api(cls, payload: Any) -> list[Self]:
        """Build a list of models from a JSON array.

        Endpoints with nothing to return sometimes answer `null` rather than an
        empty array, so that case is treated as an empty list.
        """
        if not payload:
            return []
        return [cls.from_api(item) for item in payload]

    def to_api(self) -> dict[str, Any]:
        """Render this model as the JSON object the API expects.

        Fields set to `None` are left out rather than sent as `null`. The API reads
        an absent optional field as "not supplied", which is what `None` means in
        these models: a Group carries exactly one region identifier, and the other
        three stay absent.
        """
        result: dict[str, Any] = {}
        for field in dataclasses.fields(self):
            if field.name == "extra":
                continue
            value = getattr(self, field.name)
            if value is None:
                continue
            result[camel_case(field.name)] = to_json_value(value)
        return result

File: sampleCode/R/utilities/json_helper.R
# Translation between the API's JSON and the sample's R lists.
#
# The Impact API speaks camelCase; R here is written in snake_case. These
# functions do the conversion, and `from_api()` is deliberately tolerant so the
# samples keep working when the API starts returning a field they have never
# heard of.
#
# That tolerance is the important part. The API gains fields over time, and code
# that rejects an unexpected key turns a routine addition into a broken sample,
# so unknown keys are kept rather than dropped. Print a model to see everything a
# response really contained.


# Converts an R field name to the API's spelling: hash_id -> hashId.
camel_case <- function(names) {
  vapply(
    names,
    function(name) {
      parts <- strsplit(name, "_", fixed = TRUE)[[1]]
      if (length(parts) < 2) {
        return(name)
      }
      rest <- parts[-1]
      paste0(parts[1], paste0(toupper(substring(rest, 1, 1)), substring(rest, 2), collapse = ""))
    },
    character(1),
    USE.NAMES = FALSE
  )
}


# Converts an API field name to R's spelling: hashId -> hash_id.
snake_case <- function(names) {
  tolower(gsub("([a-z0-9])([A-Z])", "\\1_\\2", names))
}


# Builds a model from one JSON object, renaming its fields and keeping the ones
# this sample does not know about.
#
# `defaults` is a named list giving the fields the model declares and the value
# each takes when the API leaves it out. Anything else the response carried is
# kept under its original name, so nothing is silently discarded.
from_api <- function(payload, defaults = list(), class_name = NULL) {
  model <- defaults

  if (!is.null(payload) && length(payload) > 0) {
    for (api_name in names(payload)) {
      value <- payload[[api_name]]
      r_name <- snake_case(api_name)
      if (r_name %in% names(defaults)) {
        # A field the model declares. JSON null arrives as NULL, which would
        # delete the entry, so keep the default in that case.
        if (!is.null(value)) {
          model[[r_name]] <- value
        }
      } else if (!is.null(value)) {
        # A field the API has added, or one this sample has no use for. Keep it
        # rather than failing.
        model[[api_name]] <- value
      }
    }
  }

  if (!is.null(class_name)) {
    class(model) <- c(class_name, "implan_model", "list")
  }
  model
}


# Builds a list of models from a JSON array.
#
# Endpoints with nothing to return sometimes answer null rather than an empty
# array, so that case becomes an empty list.
list_from_api <- function(payload, builder) {
  if (is.null(payload) || length(payload) == 0) {
    return(list())
  }
  lapply(payload, builder)
}


# Fields that must always serialize as a JSON array.
#
# The request bodies are written with `auto_unbox = TRUE`, which turns a
# length-one vector into a JSON scalar. That is what you want almost everywhere
# and wrong for these: one tag has to go out as ["capital"] and not "capital",
# and the growth report rejects a filter that is not an array. Listing them here
# rather than at each call site means a new caller cannot forget.
ARRAY_FIELDS <- c(
  "tags", "hash_ids", "urids",
  "regions", "impacts", "group_names", "event_names", "event_tags"
)


# Renders a model as the JSON object the API expects.
#
# Fields set to NULL are left out rather than sent as null. The API reads an
# absent optional field as "not supplied", which is what NULL means in these
# models: a Group carries exactly one region identifier, and the other three stay
# absent.
to_api <- function(model, array_fields = ARRAY_FIELDS) {
  payload <- list()

  for (name in names(model)) {
    value <- model[[name]]
    if (is.null(value)) {
      next
    }

    api_name <- camel_case(name)

    if (name %in% array_fields) {
      # Force an array. An unnamed list is what jsonlite renders as one, and an
      # empty one renders as [].
      payload[[api_name]] <- unname(as.list(value))
    } else if (is.list(value) && !is.null(names(value))) {
      # A nested model.
      payload[[api_name]] <- to_api(value, array_fields)
    } else if (is.list(value)) {
      # A list of nested models, such as a group's events.
      payload[[api_name]] <- unname(lapply(value, to_api, array_fields = array_fields))
    } else {
      payload[[api_name]] <- value
    }
  }

  payload
}


# One-line descriptions for the console.
#
# `describe()` is a generic so each model can say what matters about itself,
# which keeps the workflows readable: they log describe(region) rather than
# assembling the same three fields every time.
describe <- function(x, ...) UseMethod("describe")

describe.default <- function(x, ...) {
  paste(names(x), unlist(lapply(x, function(v) paste(v, collapse = "/"))),
        sep = "=", collapse = "  ")
}


# Reads a field that may be missing, which is how an optional number in a
# response is handled everywhere in these samples.
field_or <- function(model, name, default = NULL) {
  value <- model[[name]]
  if (is.null(value)) default else value
}
